### PR TITLE
Add list_file() functional API to FSSpecFileLister and IoPathFileLister

### LINKED ISSUE #387: Add functional API to FileLister
### 🚀 The feature

Another follow up on https://github.com/pytorch/data/pull/383

Since `FileLister` support `IterDataPipe` as input, we could provide functional API to it.

```py
dp = IterableWrapper([file_path_1, file_path_2, ...]).list_file()...
```
- FileLister: https://github.com/pytorch/pytorch/blob/d21154b0989c7baa058428a5d22f7b379147fec0/torch/utils/data/datapipes/iter/filelister.py#L10
- IoPathFileLister and FSSpecFileLister: https://github.com/pytorch/data/tree/main/torchdata/datapipes/iter/load

### Motivation, pitch

Better syntax

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by xiurobert**
Fixes #387

### Changes
- Adds `list_file()` method on `IoPathFileListerIterDataPipe`
- Adds `list_file()` method on `FSSpecFileListerIterDataPipe`
- Add tests for those methods

#### Additional comments
I feel as if the implementation is quite naive. Would appreciate any feedback on it.

### CODE REVIEW

**Comment by xiurobert** [comment]
Additional notes: Perhaps `return list(self)` could work as well.

**Comment by xiurobert** [comment]
@ejguan updated it along with relevant tests

File: test/test_local_io.py [hunk 1]
```diff
@@ -660,6 +660,16 @@ def test_io_path_file_lister_iterdatapipe(self):
         for path in datapipe:
             self.assertTrue(path in self.temp_sub_files)
 
+    @skipIfNoIoPath
+    def test_io_path_functional_list_files_iterdatapipe(self):
+        datapipe = IoPathFileLister(root=self.temp_sub_dir.name)
+
+        listed = list(datapipe.list_file_by_iopath())
```
Comment (ejguan) [inline test/test_local_io.py:667]: We can combine this test with the test above. And, for functional API, I suggest:
```suggestion
        datapipe = IterableWrapper([self.temp_sub_dir.name, ])
        liste = list(datapipe)
```

File: test/test_local_io.py [hunk 2]
```diff
@@ -672,6 +682,17 @@ def test_io_path_file_lister_iterdatapipe_with_list(self):
         # check all file paths within sub_folder are listed
         self.assertEqual(file_lister, all_temp_files)
 
+    @skipIfNoIoPath
+    def test_io_path_functional_list_files_iterdatapipe_with_list(self):
+        datapipe = IoPathFileLister(root=[self.temp_sub_dir.name, self.temp_sub_dir_2.name])
+
+        listed = list(datapipe.list_file_by_iopath())
```
Comment (ejguan) [inline test/test_local_io.py:689]: Ditto

File: test/test_fsspec.py [hunk 3]
```diff
@@ -82,6 +82,39 @@ def test_fsspec_file_lister_iterdatapipe_with_list(self):
         # check all file paths within sub_folder are listed
         self.assertEqual(file_lister, temp_files)
 
+    @skipIfNoFSSpec
+    def test_fsspec_functional_list_file(self):
+        datapipe = FSSpecFileLister(root="file://" + self.temp_sub_dir.name)
```
Comment (ejguan) [inline test/test_fsspec.py:87]: Ditto.

Comment (xiurobert) [reply test/test_fsspec.py]: 450ac0ed71f4d8c1cea05daccbf959febcf0b06f

File: test/test_local_io.py (hunk 2 above)
Comment (xiurobert) [reply test/test_local_io.py]: 450ac0ed71f4d8c1cea05daccbf959febcf0b06f

File: test/test_local_io.py (hunk 1 above)
Comment (xiurobert) [reply test/test_local_io.py]: 450ac0ed71f4d8c1cea05daccbf959febcf0b06f

**Comment by xiurobert** [comment]
All tests are passing on my machine so far.

File: test/test_fsspec.py [hunk 4]
```diff
@@ -64,6 +64,15 @@ def test_fsspec_file_lister_iterdatapipe(self):
                 {fsspec.implementations.local.make_path_posix(file) for file in self.temp_sub_files},
             )
 
+        # checks for functional API
+        datapipe = IterableWrapper([self.temp_sub_dir.name])
+        listed = list(datapipe.list_file_by_fsspec())
+        for path in listed:
```
Comment (ejguan) [inline test/test_fsspec.py:70]: Super nit:
```suggestion
        datapipe = datapipe.list_file_by_fsspec()
        for path in datapipe:
```

File: test/test_local_io.py [hunk 5]
```diff
@@ -660,6 +660,11 @@ def test_io_path_file_lister_iterdatapipe(self):
         for path in datapipe:
             self.assertTrue(path in self.temp_sub_files)
 
+        datapipe = IterableWrapper([self.temp_sub_dir.name])
+        listed = list(datapipe.list_file_by_iopath())
+        for path in listed:
```
Comment (ejguan) [inline test/test_local_io.py:665]: Ditto

**Review by NivekT — APPROVED** [review]
Thanks for working on this!

Question: Do you plan on working on https://github.com/pytorch/pytorch/issues/78263 as well? I see an issue but not a PR for PyTorch Core.

**Comment by xiurobert** [comment]
Yep, I do intend to do so. I opened the issue because the CONTRIBUTING.md file specifies I should be opening one before working on any "features"

File: test/test_fsspec.py (hunk 4 above)
Comment (xiurobert) [reply test/test_fsspec.py]: https://github.com/pytorch/data/pull/463/commits/c51297a767b34e04edf25534f27ac3739ab15038

Comment (ejguan) [reply test/test_fsspec.py]: Oh no, I mean you don't need to call `list` here. You can directly iterate over `datapipe`

Comment (xiurobert) [reply test/test_fsspec.py]: Ah, didn't notice. Will change it

Comment (xiurobert) [reply test/test_fsspec.py]: https://github.com/pytorch/data/pull/463/commits/a84ba1ea51a59634ffadba53e87634b3e0b4b24b

File: test/test_local_io.py (hunk 5 above)
Comment (xiurobert) [reply test/test_local_io.py]: https://github.com/pytorch/data/pull/463/commits/a84ba1ea51a59634ffadba53e87634b3e0b4b24b

**Comment by xiurobert** [comment]
@NivekT is it ok if I directly open the pull request with PyTorch core? Their contributing guidelines mention that I should be opening an issue before implementing a feature.

**Comment by xiurobert** [comment]
> Thanks for working on this!
> 
> Question: Do you plan on working on [pytorch/pytorch#78263](https://github.com/pytorch/pytorch/issues/78263) as well? I see an issue but not a PR for PyTorch Core.

Just saw the PyTorch FAQ, I could probably open the PR since it's a small change.

File: torchdata/datapipes/iter/load/fsspec.py [hunk 6]
```diff
@@ -33,6 +33,7 @@ def _assert_fsspec() -> None:
         )
 
 
+@functional_datapipe("list_file_by_fsspec")
```
Comment (NivekT) [inline torchdata/datapipes/iter/load/fsspec.py:36]:
```suggestion
@functional_datapipe("list_files_by_fsspec")
```
Based on the discussion elsewhere, let's add a 's'. Thank you so much and sorry for the inconvenience.

**Review by NivekT — CHANGES_REQUESTED** [review]

File: torchdata/datapipes/iter/load/fsspec.py (hunk 6 above)
Comment (NivekT) [reply torchdata/datapipes/iter/load/fsspec.py:36]: Note that, you will have to replace all occurrence of this usage in the PR

**Review by ejguan — APPROVED** [review]
Thank you, LGTM

**Comment by xiurobert** [comment]
Will the APIs in torchdata be updated with the new grammar as well?

**Comment by ejguan** [comment]
> Will the APIs in torchdata be updated with the new grammar as well?

Yeah. @NivekT Just opened a PR to change the names of DataPipe. See: https://github.com/pytorch/data/pull/479

**Comment by xiurobert** [comment]
Awesome.

File: test/test_local_io.py [hunk 7]
```diff
@@ -672,6 +677,11 @@ def test_io_path_file_lister_iterdatapipe_with_list(self):
         # check all file paths within sub_folder are listed
         self.assertEqual(file_lister, all_temp_files)
 
+        datapipe = IterableWrapper([self.temp_sub_dir.name, self.temp_sub_dir_2.name])
+        datapipe = datapipe.list_files_by_iopath()
+        datapipe.sort()
+        self.assertEqual(datapipe, all_temp_files)
```
Comment (ejguan) [inline test/test_local_io.py:683]: Same Error as I pointed out before.

File: test/test_fsspec.py [hunk 8]
```diff
@@ -64,6 +64,15 @@ def test_fsspec_file_lister_iterdatapipe(self):
                 {fsspec.implementations.local.make_path_posix(file) for file in self.temp_sub_files},
             )
 
+        # checks for functional API
+        datapipe = IterableWrapper([self.temp_sub_dir.name])
```
Comment (ejguan) [inline test/test_fsspec.py:68]:
```suggestion
        datapipe = IterableWrapper(["file://" + self.temp_sub_dir.name])
```

File: test/test_fsspec.py [hunk 9]
```diff
@@ -82,6 +91,12 @@ def test_fsspec_file_lister_iterdatapipe_with_list(self):
         # check all file paths within sub_folder are listed
         self.assertEqual(file_lister, temp_files)
 
+        # checks for functional API
+        datapipe = IterableWrapper([self.temp_sub_dir.name, self.temp_sub_dir_2.name])
```
Comment (ejguan) [inline test/test_fsspec.py:95]:
```suggestion
        datapipe = IterableWrapper(["file://" + self.temp_sub_dir.name, "file://" + self.temp_sub_dir_2.name])```

**Review by ejguan — CHANGES_REQUESTED** [review]
There are a few bugs in your tests.

File: test/test_fsspec.py (hunk 9 above)
Comment (xiurobert) [reply test/test_fsspec.py]: Latest commits

**Review by ejguan — COMMENTED** [review]
Could you please also add tests for `list_files_by_iopath` in https://github.com/pytorch/data/blob/aab67f314d9c8329ea688e5841abc16cbe2c4f42/test/test_local_io.py#L655-L673?

**Comment by ejguan** [comment]
You can run your test on your local machine via `python test/test_fsspec.py` and `python test/test_local_io.py`. You might need to install `fsspec` and `iopath` via `pip install fsspec iopath`.

**Comment by xiurobert** [comment]
Changes made @ejguan

**Review by ejguan — APPROVED** [review]
Thank you, LGTM

**Review by NivekT — APPROVED** [review]

**Comment by ejguan** [comment]
@xiurobert Could you please rebase this PR onto main branch?

**Comment by xiurobert** [comment]
Will do

**Comment by xiurobert** [comment]
Done. For some reason, github did not let me squash my commits.

**Comment by xiurobert** [comment]
Did this work?

**Comment by ejguan** [comment]
> Did this work?

It works now. Let me import your PR. Don't worry about squashing the commits. We will do it automatically.

**Comment by xiurobert** [comment]
Great.

**Pull request closed without merging**